File: 11-1.py
#!/usr/bin/env python

import math


class Stones:
    def __init__(self, line):
        self.stones = [int(s) for s in line.split()]

    def blink(self):
        new_stones = []
        for stone in self.stones:
            if stone == 0:
                new_stones.append(1)
            elif (math.floor(math.log10(stone)) + 1) % 2 == 0:
                engraving = str(stone)
                half = int((math.floor(math.log10(stone)) + 1) / 2)
                new_stones.append(int(engraving[:half]))
                new_stones.append(int(engraving[half:]))
            else:
                new_stones.append(stone * 2024)

        self.stones = new_stones

    def score(self):
        return len(self.stones)


stones = Stones(open('11.input').read().strip())

for _ in range(25):
    stones.blink()

print(stones.score())

File: 11-1.rb
#!/usr/bin/env ruby

class Stones
  def initialize(line)
    @stones = line.split.map(&:to_i)
  end

  def blink!
    new_stones = []
    @stones.each do |stone|
      if stone.zero?
        new_stones << 1
      elsif (Math.log10(stone).floor + 1).even?
        engraving = stone.to_s
        half = (Math.log10(stone).floor + 1) / 2
        new_stones << engraving[..half - 1].to_i
        new_stones << engraving[half..].to_i
      else
        new_stones << stone * 2024
      end
    end
    @stones = new_stones
  end

  def score
    @stones.size
  end

  def inspect
    to_s
  end

  def to_s
    "<#{self.class}: #{@stones}>"
  end
end

stones = Stones.new File.read('11.input').strip

25.times { stones.blink! }

puts stones.score
